miden-lib/asm/sat/internal/account_vault.masm

```diff
@@ -63,3 +66,227 @@ export.has_non_fungible_asset
     movdn.4 dropw movdn.4 dropw
     # => [has_asset]
 end
+
+# ADD ASSET
+# =================================================================================================
+
+# TODO: Replace with an implementation that uses the advice provider to give new amount for
+#       insertion. Ideally we shouldn't be putting vault logic in miden-vm so how can we do
+#       this ergonomically?
+
```

Comment:
One potential way to do this is to add `adv.smt_peek` decorator (this would complement the existing `adv.smt_get` and `adv.smt_insert` decorators). The decorator could work as follows:
```
Inputs:
  Operand stack: [KEY, ROOT, ...]
  Advice stack:  [...]

Outputs:
  Operand stack: [KEY, ROOT, ...]
  Advice stack:  [VALUE, ...]
```
Where `VALUE` is the value associated with the `KEY` in the TSMT with the specified `ROOT`.

We could then compute the total asset, insert it into the TSMT, and after `exec.smt::insert` check that the value on the top of the stack is the same as `VALUE`.

Reply:
Good idea! I think this provides the right degree of generality but is also well suited to our problem.  I'll create an issue for this.